samples/navigate-route/src/main/java/com/esri/arcgismaps/sample/navigateroute/components/NavigateRouteViewModel.kt

```diff
@@ -0,0 +1,417 @@
+                isTextToSpeechInitialized.set(true)
+            }
+        }
+
+        viewModelScope.launch(Dispatchers.IO) {
+            // load and set the route parameters
+            val routeParameters = routeTask.createDefaultParameters().getOrElse {
+                return@launch messageDialogVM.showMessageDialog(
+                    title = "Error creating default parameters",
+                    description = it.message.toString()
+                )
+            }.apply {
+                setStops(routeStops)
+                returnDirections = true
+                returnStops = true
+                returnRoutes = true
+            }
+
+            // get the solved route result
+            routeResult = routeTask.solveRoute(routeParameters).getOrElse {
+                return@launch messageDialogVM.showMessageDialog(
+                    title = "Error solving route",
+                    description = it.message.toString()
+                )
+            }
+
+            // reset navigation to initial state
+            resetNavigation()
+        }
+
+    }
+
+    /**
+     * Start the navigation along the provided route using the [routeResult] and
+     * collects updates in the location using the MapView's location display.
+     * */
+    fun startNavigation() {
+        val routeResult = routeResult
+            ?: return messageDialogVM.showMessageDialog("Error retrieving route result")
+        // get the route's geometry from the route result
+        val routeGeometry: Polyline = routeResult.routes[0].routeGeometry
+            ?: return messageDialogVM.showMessageDialog("Route is missing geometry")
+
+        // set up a simulated location data source which simulates movement along the route
+        val simulationParameters = SimulationParameters(
+            Instant.now(),
+            velocity = 35.0,
+            horizontalAccuracy = 5.0,
+            verticalAccuracy = 5.0
+        )
+
+        // create the simulated data source using the geometry and parameters
+        val simulatedLocationDataSource = SimulatedLocationDataSource(
+            polyline = routeGeometry,
+            parameters = simulationParameters
+        )
+
+        // set up a RouteTracker for navigation along the calculated route
+        val routeTracker = RouteTracker(
+            routeResult = routeResult,
+            routeIndex = 0,
+            skipCoincidentStops = true
+        ).apply {
+            setSpeechEngineReadyCallback {
+                isTextToSpeechInitialized.get() && textToSpeech?.isSpeaking == false
+            }
+        }
+
+
+        // create a route tracker location data source to snap the location display to the route
+        val routeTrackerLocationDataSource = RouteTrackerLocationDataSource(
+            routeTracker = routeTracker,
+            locationDataSource = simulatedLocationDataSource
+        )
+
+        locationDisplayJob = with(viewModelScope) {
+            launch(Dispatchers.IO) {
```

Comment:
```suggestion
            launch {
```

These feel like they could just run on default dispatcher. These aren't big IO pieces of work

Same as below